provider/gateway/server.go

```diff
@@ -2,19 +2,105 @@ package gateway
 
 import (
 	"context"
+	"crypto/tls"
+	"crypto/x509"
 	"net"
 	"net/http"
+	"time"
 
-	"github.com/ovrclk/akash/provider"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/pkg/errors"
 	"github.com/tendermint/tendermint/libs/log"
+
+	"github.com/ovrclk/akash/provider"
+	ctypes "github.com/ovrclk/akash/x/cert/types"
 )
 
-func NewServer(ctx context.Context, log log.Logger, pclient provider.Client, address string) *http.Server {
-	return &http.Server{
+func NewServer(
+	ctx context.Context,
+	log log.Logger,
+	pclient provider.Client,
+	cquery ctypes.QueryClient,
+	address string,
+	pid sdk.Address,
+	certs []tls.Certificate) *http.Server {
+
+	srv := &http.Server{
 		Addr:    address,
-		Handler: newRouter(log, pclient),
+		Handler: newRouter(log, pid, pclient),
 		BaseContext: func(_ net.Listener) context.Context {
 			return ctx
 		},
 	}
+
+	srv.TLSConfig = &tls.Config{
+		Certificates:       certs,
+		ClientAuth:         tls.RequestClientCert,
+		InsecureSkipVerify: true,
+		MinVersion:         tls.VersionTLS13,
+		VerifyPeerCertificate: func(certificates [][]byte, _ [][]*x509.Certificate) error {
```

Comment:
This seems very similar to the another method named `verifyPeerCertificate`. Can we reduce the duplication?

Reply:
No, they work differently. Client certificate not always supplied